test/varmat_compatibility.py

```diff
@@ -0,0 +1,260 @@
+TEST_FOLDER = os.path.abspath(os.path.join(HERE, "..", "test"))
+sys.path.append(TEST_FOLDER)
+WORKING_FOLDER = "test/varmat-compatibility"
+
+BENCHMARK_TEMPLATE = """
+static void {benchmark_name}() {{
+{code}
+}}
+"""
+
+def run_command(command):
+    """
+    Runs given command and waits until it finishes executing.
+    :param command: command to execute
+    """
+    proc = subprocess.Popen(command, stdout = subprocess.PIPE, stderr = subprocess.PIPE)
+    stdout, stderr = proc.communicate()
+    
+    if proc.poll() == 0:
+        return (True, stdout, stderr)
+    else:
+        return (False, stdout, stderr)
+
+def build_signature(cpp_code):
+    """
+    Try to build the given cpp code
+
+    Return a tuple (successful, attempted)
+
+    attempted is true if there was an attempt to build the code
+
+    successful is true is the code was successfully built
+
+    :param cpp_code Code to build
+    """
+    attempted = False
+    successful = False
+
+    if cpp_code:
+        attempted = True
+
+        f = tempfile.NamedTemporaryFile("w", dir = WORKING_FOLDER, suffix = "_test.cpp", delete = False)
+        f.write("#include <test/expressions/expression_test_helpers.hpp>\n\n")
+        f.write(cpp_code)
+        f.close()
+
+        cpp_path = os.path.join(WORKING_FOLDER, os.path.basename(f.name))
+
+        object_path = cpp_path.replace(".cpp", ".o")
+        dependency_path = cpp_path.replace(".cpp", ".d")
+        stdout_path = cpp_path.replace(".cpp", ".stdout")
+        stderr_path = cpp_path.replace(".cpp", ".stderr")
+
+        successful, stdout, stderr = run_command([make, object_path])
+
+        # Only clean up files if check succesful
+        if successful:
+            try:
+                os.remove(cpp_path)
+                os.remove(dependency_path)
+                os.remove(object_path)
+            except OSError:
+                pass
+        else:
+            with open(stdout_path, "w") as stdout_f:
+                stdout_f.write(stdout.decode("utf-8"))
+
+            with open(stderr_path, "w") as stderr_f:
+                stderr_f.write(stderr.decode("utf-8"))
+        
+    return successful, attempted
+
+def main(functions_or_sigs, results_file, cores):
+    """
+    Attempt to build all the signatures in functions_or_sigs, or all the signatures
+    associated with all the functions in functions_or_sigs, or if functions_or_sigs
+    is empty every signature the stanc3 compiler exposes.
+
+    Results are written to a results json file. Individual signatures are classified
+    as either compatible, incompatible, or impossible
+
+    Compatible signatures can be compiled with varmat types in every argument that
+    could possibly be a varmat (the matrix-like ones).
+
+    Incompatible signatures cannot all be built, and for impossible signatures it does
+    not make sense to try to build them (there are no matrix arguments, or the function
+    does not support reverse mode autodiff, etc)
```

Comment:
I like irrelevant. Can switch.

But this also feels it could be a game to come up with as many `i` words that mean a similar thing. I'll counter first with `inconsquential`.

Reply:
Right, none of these words perfectly describe it. But I still like irrelevant bets among what was suggested. Lets consider this an optional change.

Reply:
Nah nah, I genuinely like irrelevant.

What I said came out kinda toxic my bad. I just realized I had a couple other words that start with `i` lol (also immaterial).